03_letters/CameronSchafer/main_test.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+// main() test
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+	expected := `a:2
+b:1
+`
+	actual := buf.String()
+
+	assert.Equalf(t, expected, actual,
+		"Unexpected output in main()\nexpected: %q\nactual: %q", expected, actual)
+}
+
+// sortLetters()
+func TestSortLetters(t *testing.T) {
+	// test cases with descriptions.
+	testCases := map[string]struct {
+		input       map[rune]int
+		expectedArr []string
+	}{
+		"2 letters singles [a:1, b:1]": {input: map[rune]int{'a': 1, 'b': 1},
+			expectedArr: []string{"a:1", "b:1"},
+		},
+		"2 letters multiples": {input: map[rune]int{'a': 7, 'b': 7},
+			expectedArr: []string{"a:7", "b:7"},
+		},
+		"1 letter, 1 empty": {input: map[rune]int{' ': 2, 'a': 1},
+			expectedArr: []string{" :2", "a:1"},
+		},
+		"1 letter": {input: map[rune]int{'A': 7},
+			expectedArr: []string{"A:7"},
+		},
+		"3 letters": {input: map[rune]int{'A': 1, 'B': 1, 'c': 2},
+			expectedArr: []string{"A:1", "B:1", "c:2"},
+		},
+		"3 random order": {input: map[rune]int{'B': 1, 'A': 1, 'c': 2},
+			expectedArr: []string{"A:1", "B:1", "c:2"},
+		},
+	}
+
+	for name, test := range testCases {
+		test := test
+		name := name
+		// t.Run creates a sub test and runs it like a normal test
+		t.Run(name, func(t *testing.T) {
+			resultArr := sortLetters(test.input)
+			assert.Equalf(t, test.expectedArr, resultArr,
+				"Unexpected output in %v\nexpected: %v,\nactual: %v",
+				name, test.expectedArr, resultArr)
+
+		})
+	}
+}
+
+// letters()
+func TestLetters(t *testing.T) {
+	// test cases with descriptions.
+	testCases := map[string]struct {
+		input       string
+		expectedArr map[rune]int
+	}{
+		"3 letters": {input: "aba",
+			expectedArr: map[rune]int{'a': 2, 'b': 1},
+		},
+		"letters with spaces": {input: "a a ",
+			expectedArr: map[rune]int{' ': 2, 'a': 2},
+		},
+		"AaAa}{}{}}{}{}{": {input: "AaAa}{}{}}{}{}{",
+			expectedArr: map[rune]int{'A': 2, 'a': 2, '{': 5, '}': 6},
+		},
+		"empty string": {input: "",
+			expectedArr: map[rune]int{},
+		},
+		"only spaces": {input: "    ",
```

Comment:
You can also add a case for special characters.
```
"special characters": {
			input: "无😊👍👍",
			expectedArr:  map[rune]int{ '👍': 2, '无': 1, '😊': 1},
		}
```

Reply:
Done